test/tstQueryTreeRay.cpp

```diff
@@ -131,4 +154,54 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(test_ray_box_intersection, DeviceType,
       make_reference_solution<int>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, {0, 10}));
 }
 
+template <typename DeviceType>
+struct InsertIntersections
+{
+  mutable int _i;
+  Kokkos::View<int *, DeviceType> _ordered_intersections;
+
+  template <typename Predicate>
+  KOKKOS_FUNCTION void operator()(Predicate const &,
+                                  int const primitive_index) const
+  {
+    _ordered_intersections(_i++) = primitive_index;
+  }
+};
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(test_ray_box_intersection_new, DeviceType,
+                              ARBORX_TEST_DEVICE_TYPES)
+{
+  using memory_space = typename DeviceType::memory_space;
+  typename DeviceType::execution_space exec_space;
+
+  std::vector<ArborX::Box> boxes;
+  int const n = 10;
+  for (unsigned int i = 0; i < n; ++i)
+    boxes.emplace_back(ArborX::Point(i, i, i),
+                       ArborX::Point(i + 1, i + 1, i + 1));
+  Kokkos::View<ArborX::Box *, DeviceType> device_boxes("boxes", 10);
+  Kokkos::deep_copy(exec_space, device_boxes,
+                    Kokkos::View<ArborX::Box *, Kokkos::HostSpace>(
+                        boxes.data(), boxes.size()));
+
+  ArborX::BVH<memory_space> const tree(exec_space, device_boxes);
+
+  ArborX::Experimental::Ray ray{
+      ArborX::Point{0, 0, 0},
+      ArborX::Experimental::Vector{1. / n, 1. / n, 1. / n}};
+  Kokkos::View<ArborX::Experimental::Ray *, DeviceType> device_rays("rays", 1);
+  Kokkos::View<int *, DeviceType> device_ordered_intersections(
+      "ordered_intersections", n);
+  Kokkos::deep_copy(exec_space, device_rays, ray);
+
+  BoxesIntersectedByRayOrdered<DeviceType> predicates{device_rays};
+
+  tree.query(exec_space, predicates,
+             InsertIntersections<DeviceType>{0, device_ordered_intersections});
+  auto const host_ordered_intersections = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace{}, device_ordered_intersections);
+  for (unsigned int i = 0; i < n; ++i)
+    BOOST_TEST(host_ordered_intersections(i) == i);
+}
```

Comment:
Take advantage of our `Kokkos::View` adapters and the Boost native support to compare collection of values (grep for `per_element`)
We try to keep boiler plate code minimal to focus on the test content.

Reply:
I didn't because it's more concise this way IMHO but sure I'll change, see [03e716f](https://github.com/arborx/ArborX/pull/683/commits/03e716fd2a379b930005071710e677658ae50059)